### ISSUE TITLE
How does the accuracy and perplexity calculated while training?

### ISSUE THREAD

**Issue by prabalbansal** [OpenNMT/OpenNMT-py#1728]
It is calculated on batch level on validation set?(not sure need some more explanation)
As during my training, sometimes accuracy drops sharply then in next 50th step it comes to normal particularly in transformer model.

Could you explain why it behaves as such.

**Comment by Zenglinxiao** [comment]
> It is calculated on batch level on validation set?(not sure need some more explanation)

It is calculated on the whole validation set. We initialize a Statistic object and report its info at end of the evaluation.

> As during my training, sometimes accuracy drops sharply then in next 50th step it comes to normal particularly in transformer model.

During the training, we keep two `Statistic` object: `total_stat`, `report_stats`.
One called `total_stat` which is continuously updated after each batch and will be returned at the end of the training. This stats will NOT be print to the terminal.
The other called `report_stats`, this is what you see from the log or tensorboard. It will also be updated continuously after each batch, but will only be print to the terminal or tensorboard if the training step reaches N * `--report_every` step. Once printed out, it will be reassigned with a new `Statistic` and start a new round of calculating.

**Comment by prabalbansal** [comment]
Okay thanks for the reply. I set report_every = 50 . But why report_stats drops sharply.
![image](https://user-images.githubusercontent.com/30004110/75960098-c1b74780-5ebf-11ea-9159-a181d583f81c.png)
